docs/user_guide/decoupled_models.md

```diff
@@ -87,10 +87,100 @@ exactly one response per request. Even standard ModelInfer RPC in the GRPC endpo
 your application should be cognizant that the callback function you registered with 
 `TRITONSERVER_InferenceRequestSetResponseCallback` can be invoked any number of times,
 each time with a new response. You can take a look at [grpc_server.cc](https://github.com/triton-inference-server/server/blob/main/src/grpc_server.cc)
+
+### Knowing When a Decoupled Inference Request is Complete
+
+A request is considered complete when a response containing the
+`TRITONSERVER_RESPONSE_COMPLETE_FINAL` flag is received. For decoupled models,
+there are two ways this can happen. The model/backend calls one of the
+following [TRITONBACKEND APIs](https://github.com/triton-inference-server/core/blob/main/include/triton/core/tritonbackend.h):
+1. `TRITONBACKEND_ResponseSend(response, TRITONSERVER_RESPONSE_COMPLETE_FINAL, ...)`
+2. `TRITONBACKEND_ResponseFactorySendFlags(factory, TRITONSERVER_RESPONSE_COMPLETE_FINAL)`
+
+As described in the
+[backend repo](https://github.com/triton-inference-server/backend/blob/main/README.md#special-cases)
+for decoupled models:
+
+> If the backend should not send any more responses for the request,
+> `TRITONBACKEND_ResponseFactorySendFlags` can be used to send 
+> `TRITONSERVER_RESPONSE_COMPLETE_FINAL` using the ResponseFactory.
+
+In the `TRITONBACKEND_ResponseFactorySendFlags` case, only the `flags` are
+communicated back to the frontend to update some internal state, and there 
+is no actual Inference Response sent back along with the flags. The default
+behavior in this case is not to send anything back to the client, as there
+is no response to send.
+
+In some cases, this default behavior proved to be significantly more performant.
+For example, take a decoupled model with an `N` request -> `1` response structure.
+For each of the first `N-1` requests, the model will send "zero" responses back
+by using `TRITONBACKEND_ResponseFactorySendFlags` as described above, and likely
+update some internal states in the model. Finally, on the `N`th request, the
+model is ready to send a response. 
+
+If the client is written in such a way that it is aware of the model's
+expected behavior, it can save resources and avoid network contention by
+not needing to communicate the `N-1` "empty" responses back and forth with
+the client, and instead the client will just wait until it receives the single
+non-empty response expected at the end on request `N`.
+
+However, there are cases where a user may want to write a client that can
+generically handle any model, without knowing implementation details about it.
+Similarly, there are cases where the number of responses a model will send
+is unknown beforehand, so the client may need a programmatic way to know when
+the final response for a given request has been received. A common case for
+this may be where for a language model that has a `1` request -> `N` response
+structure. 
+
+To handle this case, Triton exposes a boolean `"triton_final_response"` response 
```

Comment:
need to add to the extension_parameters doc.

Also - maybe a line on the sequence extension doc (we reference grpc streaming mode there).

I think streaming mode should probably be its own extension doc - and ideally the new documentation would actually live there - as the changes are only for GRPC streaming mode - no changes for C API or for backend in actuality -

Reply:
1. Added mention of the reserved params in parameters doc.
2. Took a look at sequence doc, but not sure where it makes sense to add anything at the moment.
3. Agreed streaming extension could be its own doc, but will defer to a future PR. Maybe along with KServe discussions.

Lmk what you think.

Reply:
Sounds good